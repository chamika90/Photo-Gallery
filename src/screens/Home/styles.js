import { StyleSheet } from 'react-native';
import { theme } from 'config/theme';

const { colors } = theme;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
    backgroundColor: colors.primaryImageBackgroundColor,
  },
  imageListContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  imageContainer: {
    margin: 5,
    borderRadius: 10,
  },
  image: {
    borderRadius: 10,
    width: 150,
    height: 150,
    backgroundColor: colors.secondaryImageBackgroundColor,
  },
  emptyListContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  emptyListMessage: {
    color: colors.secondaryImageBackgroundColor,
  },
});

export default styles;
