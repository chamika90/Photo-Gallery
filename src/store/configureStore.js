import { createStore, applyMiddleware } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import AsyncStorage from '@react-native-community/async-storage';
import createSagaMiddleware from 'redux-saga';
import expireReducer from 'redux-persist-expire';

import rootReducers from 'redux/reducers';
import sagas from 'sagas';

const sagaMiddleware = createSagaMiddleware();

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whiteList: ['galleryReducer'],
  transforms: [
    expireReducer('galleryReducer', {
      expireSeconds: 60 * 60 * 24,
      expiredState: [],
      autoExpire: true,
    }),
  ],
};
const persistedReducer = persistReducer(persistConfig, rootReducers);

const store = createStore(persistedReducer, applyMiddleware(sagaMiddleware));

const persister = persistStore(store);
const configureStore = () => {
  return { persister, store };
};
sagaMiddleware.run(sagas);

export default configureStore;
