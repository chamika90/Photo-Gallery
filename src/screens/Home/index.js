import React, { useEffect, useCallback } from 'react';
import {
  SafeAreaView,
  StatusBar,
  TouchableOpacity,
  View,
  Text,
  FlatList,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import FastImage from 'react-native-fast-image';

import { getGalleryImages } from 'redux/actions/photoGalleryActions';
import { theme } from 'config/theme';
import styles from './styles';

const { colors } = theme;

//EmptyList Component
const EmptyList = () => {
  return (
    <View style={styles.emptyListContainer} testID="emptyView">
      <Text style={styles.emptyListMessage}>No Images Found</Text>
    </View>
  );
};

// Photo Card Component
const PhotoCard = ({ photo, onSelectPhoto }) => {
  console.log('photo card');
  return (
    <TouchableOpacity
      style={styles.imageContainer}
      onPress={() => onSelectPhoto(photo)}>
      <FastImage
        style={styles.image}
        source={{
          uri: photo.download_url,
          cache: FastImage.cacheControl.immutable,
          priority: FastImage.priority.normal,
        }}
      />
    </TouchableOpacity>
  );
};

/*
 * Home Screen
 */
const Home = ({ navigation }) => {
  const pageSize = 12;
  const dispatch = useDispatch();
  const imageList = useSelector(state => state.galleryReducer.photoList);
  const page = useSelector(state => state.galleryReducer.page);
  const hasNextPage = useSelector(state => state.galleryReducer.hasNextPage);

  useEffect(() => {
    if (!imageList) {
      getImageList();
    }
  }, [imageList, getImageList]);

  const getImageList = useCallback(() => {
    const payload = {
      page: page + 1,
      limit: pageSize,
    };
    dispatch(getGalleryImages(payload));
  }, [dispatch, page]);

  const handlePhotoSelect = photo => {
    navigation.navigate('PhotoView', {
      item: photo,
      itemList: imageList,
    });
  };

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar
        backgroundColor={colors.primaryHeaderBackgroundColor}
        barStyle="light-content"
      />

      <FlatList
        testID="photoList"
        contentContainerStyle={styles.imageListContainer}
        numColumns={2}
        data={imageList}
        renderItem={({ item, index, separators }) => (
          <PhotoCard
            testID="photoCard"
            key={item.id}
            photo={item}
            onSelectPhoto={photo => {
              handlePhotoSelect(photo);
            }}
          />
        )}
        ListEmptyComponent={<EmptyList />}
        onEndReached={() => {
          if (hasNextPage) {
            getImageList();
          }
        }}
      />
    </SafeAreaView>
  );
};

export default Home;
