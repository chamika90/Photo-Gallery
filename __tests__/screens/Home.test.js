import 'react-native';
import React from 'react';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {render} from '@testing-library/react-native';

import Home from 'screens/Home';

describe('Home', () => {
  const initialState = {
    galleryReducer: {
      photoList: [
        {
          author: 'Alejandro Escamilla',
          download_url: 'https://picsum.photos/id/1/5616/3744',
          height: 3744,
          id: '1',
          url: 'https://unsplash.com/photos/LNRyGwIJr5c',
          width: 5616,
        },
        {
          author: 'Paul Jarvis',
          download_url: 'https://picsum.photos/id/10/2500/1667',
          height: 1667,
          id: '10',
          url: 'https://unsplash.com/photos/6J--NXulQCs',
          width: 2500,
        },
        {
          author: 'Tina Rataj',
          download_url: 'https://picsum.photos/id/100/2500/1656',
          height: 1656,
          id: '100',
          url: 'https://unsplash.com/photos/pwaaqfoMibI',
          width: 2500,
        },
      ],
      page: 0,
      hasNextPage: true,
    },
  };
  const mockStore = configureStore();
  let store;

  it('should match snapshot', async () => {
    store = mockStore(initialState);
    const {toJSON} = render(
      <Provider store={store}>
        <Home />,
      </Provider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
