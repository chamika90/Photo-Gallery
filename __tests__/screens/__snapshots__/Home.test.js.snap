// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#000000",
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "center",
        }
      }
      onScroll={[Function]}
    >
      <View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderRadius": 10,
              "margin": 5,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 10,
                  "height": 150,
                  "width": 150,
                },
              ]
            }
          >
            <FastImageView
              resizeMode="cover"
              source={
                Object {
                  "cache": "immutable",
                  "priority": "normal",
                  "uri": "https://picsum.photos/id/1/5616/3744",
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderRadius": 10,
              "margin": 5,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 10,
                  "height": 150,
                  "width": 150,
                },
              ]
            }
          >
            <FastImageView
              resizeMode="cover"
              source={
                Object {
                  "cache": "immutable",
                  "priority": "normal",
                  "uri": "https://picsum.photos/id/10/2500/1667",
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderRadius": 10,
              "margin": 5,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 10,
                  "height": 150,
                  "width": 150,
                },
              ]
            }
          >
            <FastImageView
              resizeMode="cover"
              source={
                Object {
                  "cache": "immutable",
                  "priority": "normal",
                  "uri": "https://picsum.photos/id/100/2500/1656",
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>,
  ",",
]
`;
